File: packages/extract-py/src/synapt_extract/finalize.py
"""Three-stage finalization pipeline for SynaptExtraction IL v1."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

from synapt_extract.validate import ValidationResult, validate_extraction


@dataclass
class FinalizeContext:
    produced_by: str
    user_id: str | None = None
    source_id: str | None = None
    source_type: str | None = None
    kind: str | None = None
    extensions: dict[str, Any] | None = None
    embeddings: list[dict[str, Any]] | None = None
    capabilities_hint: list[str] | None = None


@dataclass
class FinalizeResult:
    extraction: dict[str, Any]
    validation: ValidationResult
    warnings: list[str] = field(default_factory=list)


def _has_payload_beyond_version(obj: dict[str, Any]) -> bool:
    return any(k != "version" for k in obj)


def _detect_capabilities(doc: dict[str, Any]) -> list[str]:
    caps: list[str] = []
    entities = doc.get("entities", [])
    goals = doc.get("goals", [])

    if isinstance(entities, list) and entities:
        caps.append("entities")
        if any(e.get("state") is not None for e in entities):
            caps.append("entity_state")
        if any(e.get("context") is not None or e.get("date_hint") is not None for e in entities):
            caps.append("entity_context")
        if any(e.get("id") is not None for e in entities):
            caps.append("entity_ids")
        all_rels = [r for e in entities for r in (e.get("relations") or [])]
        if all_rels:
            caps.append("relations")
            if any(r.get("origin") is not None for r in all_rels):
                caps.append("relation_origin")
        if any(e.get("source") is not None for e in entities):
            caps.append("evidence_anchoring")
        if any(e.get("signals") is not None for e in entities):
            caps.append("assertion_signals")

    if isinstance(goals, list) and goals:
        caps.append("goals")
        if any(g.get("stated_at") is not None or g.get("resolved_at") is not None for g in goals):
            caps.append("goal_timing")
        if any(isinstance(g.get("entity_refs"), list) and g["entity_refs"] for g in goals):
            caps.append("goal_entity_refs")
        if "evidence_anchoring" not in caps and any(g.get("source") is not None for g in goals):
            caps.append("evidence_anchoring")
        if "assertion_signals" not in caps and any(g.get("signals") is not None for g in goals):
            caps.append("assertion_signals")

    themes = doc.get("themes")
    if isinstance(themes, list) and themes:
        caps.append("themes")
    if isinstance(doc.get("summary"), str):
        caps.append("summary")
    if isinstance(doc.get("sentiment"), str):
        caps.append("sentiment")

    facts = doc.get("facts", [])
    if isinstance(facts, list) and facts:
        caps.append("facts")
        if "evidence_anchoring" not in caps and any(f.get("source") is not None for f in facts):
            caps.append("evidence_anchoring")
        if "assertion_signals" not in caps and any(f.get("signals") is not None for f in facts):
            caps.append("assertion_signals")

    temporal = doc.get("temporal_refs", [])
    if isinstance(temporal, list) and temporal:
        caps.append("temporal_refs")
        if any(r.get("type") is not None or r.get("resolved_end") is not None for r in temporal):
            caps.append("temporal_classes")

    return caps


def _inject_sub_versions(obj: dict[str, Any]) -> None:
    obj["version"] = "1"


def finalize_extraction(
    llm_output: dict[str, Any],
    context: FinalizeContext,
) -> FinalizeResult:
    """Assemble a complete SynaptExtraction from LLM output and client context.

    Stage 1 (llm_output): content fields from the LLM.
    Stage 2: injects client context (produced_by, user_id, embeddings, etc.).
    Stage 3: injects sub-schema versions, computes capabilities, validates.
    """
    warnings: list[str] = []
    doc = dict(llm_output)

    # Stage 2: inject client context
    doc["version"] = "1"
    doc["produced_by"] = context.produced_by
    if context.user_id is not None:
        doc["user_id"] = context.user_id
    if context.source_id is not None:
        doc["source_id"] = context.source_id
    if context.source_type is not None:
        doc["source_type"] = context.source_type
    if context.kind is not None:
        doc["kind"] = context.kind
    if context.extensions is not None:
        doc["extensions"] = context.extensions

    if context.embeddings is not None:
        finalized_embs = []
        for emb in context.embeddings:
            full = {"version": "1", **emb}
            if "dimensions" not in full and isinstance(full.get("vector"), list):
                full["dimensions"] = len(full["vector"])
            finalized_embs.append(full)
        doc["embeddings"] = finalized_embs

    # Stage 3: inject sub-schema versions, strip empty sub-schemas
    for items_key in ("entities", "goals", "facts"):
        items = doc.get(items_key)
        if not isinstance(items, list):
            continue
        for item in items:
            if not isinstance(item, dict):
                continue
            for sub_key in ("source", "signals"):
                sub = item.get(sub_key)
                if isinstance(sub, dict):
                    if _has_payload_beyond_version(sub):
                        _inject_sub_versions(sub)
                    else:
                        del item[sub_key]
            if items_key == "entities" and isinstance(item.get("relations"), list):
                for rel in item["relations"]:
                    if isinstance(rel, dict) and isinstance(rel.get("signals"), dict):
                        sig = rel["signals"]
                        if _has_payload_beyond_version(sig):
                            _inject_sub_versions(sig)
                        else:
                            del rel["signals"]

    temporal = doc.get("temporal_refs")
    if isinstance(temporal, list):
        for ref in temporal:
            if isinstance(ref, dict):
                _inject_sub_versions(ref)

    # Stage 3: compute capabilities from observed payload
    observed = _detect_capabilities(doc)
    if context.capabilities_hint:
        for hinted in context.capabilities_hint:
            if hinted not in observed:
                warnings.append(
                    f'capabilities_hint includes "{hinted}" but payload does not contain it; using observed'
                )
    doc["capabilities"] = observed

    # Stage 3: capability implication warnings
    if "goal_entity_refs" in observed and "entity_ids" not in observed:
        warnings.append(
            "goal_entity_refs present but entity_ids missing; "
            "entity ref resolution will fall back to name matching"
        )

    # Stage 3: validate
    validation = validate_extraction(doc)

    return FinalizeResult(
        extraction=doc,
        validation=validation,
        warnings=warnings,
    )

File: packages/extract/src/finalize.ts
import type {
  SynaptExtraction,
  SynaptEntity,
  SynaptGoal,
  SynaptFact,
  SynaptEmbedding,
  SynaptTemporalRef,
  ExtractionCapability,
} from "./schema.js";
import { validateExtraction, type ValidationResult } from "./validate.js";

export interface FinalizeContext {
  produced_by: string;
  user_id?: string;
  source_id?: string;
  source_type?: string;
  kind?: string;
  extensions?: Record<string, unknown>;
  embeddings?: Omit<SynaptEmbedding, "version">[];
  capabilities_hint?: ExtractionCapability[];
}

export interface FinalizeResult {
  extraction: SynaptExtraction;
  validation: ValidationResult;
  warnings: string[];
}

const CAPABILITY_DEPS: Record<string, string[]> = {
  entity_state: ["entities"],
  entity_context: ["entities"],
  entity_ids: ["entities"],
  goal_timing: ["goals"],
  goal_entity_refs: ["goals", "entity_ids"],
  temporal_classes: ["temporal_refs"],
  relations: ["entities", "entity_ids"],
  relation_origin: ["relations"],
};

function injectSubSchemaVersions(obj: Record<string, unknown>): void {
  if (obj && typeof obj === "object") {
    obj.version = "1";
  }
}

function hasPayloadBeyondVersion(obj: Record<string, unknown>): boolean {
  return Object.keys(obj).some((k) => k !== "version");
}

function detectCapabilities(doc: Record<string, unknown>): ExtractionCapability[] {
  const caps: ExtractionCapability[] = [];
  const entities = doc.entities as Record<string, unknown>[] | undefined;
  const goals = doc.goals as Record<string, unknown>[] | undefined;

  if (Array.isArray(entities) && entities.length > 0) {
    caps.push("entities");
    if (entities.some((e) => e.state !== undefined)) caps.push("entity_state");
    if (entities.some((e) => e.context !== undefined || e.date_hint !== undefined)) caps.push("entity_context");
    if (entities.some((e) => e.id !== undefined)) caps.push("entity_ids");
    if (entities.some((e) => Array.isArray(e.relations) && (e.relations as unknown[]).length > 0)) {
      caps.push("relations");
      const allRelations = entities.flatMap((e) => (e.relations as Record<string, unknown>[]) || []);
      if (allRelations.some((r) => r.origin !== undefined)) caps.push("relation_origin");
    }
    if (entities.some((e) => e.source !== undefined)) caps.push("evidence_anchoring");
    if (entities.some((e) => e.signals !== undefined)) caps.push("assertion_signals");
  }

  if (Array.isArray(goals) && goals.length > 0) {
    caps.push("goals");
    if (goals.some((g) => g.stated_at !== undefined || g.resolved_at !== undefined)) caps.push("goal_timing");
    if (goals.some((g) => Array.isArray(g.entity_refs) && (g.entity_refs as unknown[]).length > 0)) {
      caps.push("goal_entity_refs");
    }
    if (!caps.includes("evidence_anchoring") && goals.some((g) => g.source !== undefined)) {
      caps.push("evidence_anchoring");
    }
    if (!caps.includes("assertion_signals") && goals.some((g) => g.signals !== undefined)) {
      caps.push("assertion_signals");
    }
  }

  if (Array.isArray(doc.themes) && (doc.themes as unknown[]).length > 0) caps.push("themes");
  if (typeof doc.summary === "string") caps.push("summary");
  if (typeof doc.sentiment === "string") caps.push("sentiment");

  if (Array.isArray(doc.facts) && (doc.facts as unknown[]).length > 0) {
    caps.push("facts");
    const facts = doc.facts as Record<string, unknown>[];
    if (!caps.includes("evidence_anchoring") && facts.some((f) => f.source !== undefined)) {
      caps.push("evidence_anchoring");
    }
    if (!caps.includes("assertion_signals") && facts.some((f) => f.signals !== undefined)) {
      caps.push("assertion_signals");
    }
  }

  if (Array.isArray(doc.temporal_refs) && (doc.temporal_refs as unknown[]).length > 0) {
    caps.push("temporal_refs");
    const refs = doc.temporal_refs as Record<string, unknown>[];
    if (refs.some((r) => r.type !== undefined || r.resolved_end !== undefined)) caps.push("temporal_classes");
  }

  return caps;
}

export function finalizeExtraction(
  llmOutput: Record<string, unknown>,
  context: FinalizeContext,
): FinalizeResult {
  const warnings: string[] = [];
  const doc = { ...llmOutput };

  // Stage 2: inject client context
  doc.version = "1";
  doc.produced_by = context.produced_by;
  if (context.user_id !== undefined) doc.user_id = context.user_id;
  if (context.source_id !== undefined) doc.source_id = context.source_id;
  if (context.source_type !== undefined) doc.source_type = context.source_type;
  if (context.kind !== undefined) doc.kind = context.kind;
  if (context.extensions !== undefined) doc.extensions = context.extensions;

  // Stage 2: inject embeddings with version
  if (context.embeddings !== undefined) {
    doc.embeddings = context.embeddings.map((emb) => {
      const full: Record<string, unknown> = { version: "1", ...emb };
      if (full.dimensions === undefined && Array.isArray(full.vector)) {
        full.dimensions = (full.vector as number[]).length;
      }
      return full;
    });
  }

  // Stage 3: inject sub-schema versions on source refs and signals
  if (Array.isArray(doc.entities)) {
    for (const ent of doc.entities as Record<string, unknown>[]) {
      if (ent.source && typeof ent.source === "object") {
        const src = ent.source as Record<string, unknown>;
        if (hasPayloadBeyondVersion(src)) {
          injectSubSchemaVersions(src);
        } else {
          delete ent.source;
        }
      }
      if (ent.signals && typeof ent.signals === "object") {
        const sig = ent.signals as Record<string, unknown>;
        if (hasPayloadBeyondVersion(sig)) {
          injectSubSchemaVersions(sig);
        } else {
          delete ent.signals;
        }
      }
      if (Array.isArray(ent.relations)) {
        for (const rel of ent.relations as Record<string, unknown>[]) {
          if (rel.signals && typeof rel.signals === "object") {
            const sig = rel.signals as Record<string, unknown>;
            if (hasPayloadBeyondVersion(sig)) {
              injectSubSchemaVersions(sig);
            } else {
              delete rel.signals;
            }
          }
        }
      }
    }
  }

  if (Array.isArray(doc.goals)) {
    for (const goal of doc.goals as Record<string, unknown>[]) {
      if (goal.source && typeof goal.source === "object") {
        const src = goal.source as Record<string, unknown>;
        if (hasPayloadBeyondVersion(src)) {
          injectSubSchemaVersions(src);
        } else {
          delete goal.source;
        }
      }
      if (goal.signals && typeof goal.signals === "object") {
        const sig = goal.signals as Record<string, unknown>;
        if (hasPayloadBeyondVersion(sig)) {
          injectSubSchemaVersions(sig);
        } else {
          delete goal.signals;
        }
      }
    }
  }

  if (Array.isArray(doc.facts)) {
    for (const fact of doc.facts as Record<string, unknown>[]) {
      if (fact.source && typeof fact.source === "object") {
        const src = fact.source as Record<string, unknown>;
        if (hasPayloadBeyondVersion(src)) {
          injectSubSchemaVersions(src);
        } else {
          delete fact.source;
        }
      }
      if (fact.signals && typeof fact.signals === "object") {
        const sig = fact.signals as Record<string, unknown>;
        if (hasPayloadBeyondVersion(sig)) {
          injectSubSchemaVersions(sig);
        } else {
          delete fact.signals;
        }
      }
    }
  }

  if (Array.isArray(doc.temporal_refs)) {
    for (const ref of doc.temporal_refs as Record<string, unknown>[]) {
      injectSubSchemaVersions(ref);
    }
  }

  // Stage 3: compute capabilities from observed payload
  const observed = detectCapabilities(doc);
  if (context.capabilities_hint) {
    for (const hinted of context.capabilities_hint) {
      if (!observed.includes(hinted)) {
        warnings.push(`capabilities_hint includes "${hinted}" but payload does not contain it; using observed`);
      }
    }
  }
  doc.capabilities = observed;

  // Stage 3: capability implication enforcement
  if (observed.includes("goal_entity_refs") && !observed.includes("entity_ids")) {
    warnings.push("goal_entity_refs present but entity_ids missing; entity ref resolution will fall back to name matching");
  }

  // Stage 3: validate
  const validation = validateExtraction(doc);

  return {
    extraction: doc as unknown as SynaptExtraction,
    validation,
    warnings,
  };
}
